### PR TITLE
Spyre-cli: Add direct launcher for spyrecode

### PR DESCRIPTION
**Pull request by ChanderG**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [X] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

Provides a new tool which allows to directly run pre-compiled spyreCodeDir on real Spyre hardware relying on torch-spyre for the Runtime software only (no compile time pathway of torch-spyre is used here).

Initial version, with a few more planned features pending. This version is tested to work with spyreCode created from SDSC files (ie the current torch-spyre pathway), but not yet tested with KTIR etc.

### CODE REVIEW

**Review by ashokponkumar — DISMISSED** [review]

**Review by dushyantbehl — APPROVED** [review]
Looks good to me.

File: extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
+  "inputs": [
+    {
+      "ndims": 2,
+      "dtype": "float16"
```
Comment (fabianlim) [inline extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json:4]: what happens in the future when the dynamic shapes is supposed, and some of the dimensinos are wildcards,

File: extensions/spyre-cli/spyre_cli/core.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    try:
+        spec = parse_json_file(iofile)
+    except Exception as e:
+        print(e)
+        exit(1)
+
+    import torch
+
+    tensors = []
+    for tensor_spec in spec.inputs + spec.outputs:
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/core.py]: The tool binds inputs + outputs positionally, and SpyreStream::launch only checks is_privateuseone() — not count, order, or dtype. Since the compiled bundle is the source of truth for arg identity/order, io.json is effectively an unchecked second copy of that contract: a mismatch binds tensors to the wrong segments and produces silent garbage rather than an error.

Can we add at least a count check — len(tensors) against the external-I/O arg count the prepared jobplan expects? It's cheap, needs no extra metadata, and catches the worst case. Ideally we'd also validate dtype/order/shape against the bundle's own signature so io.json can't drift from it. The README documents this foot-gun (constraint #2), which is good, but docs alone leave the silent-failure path in place — I'd like at least the count check before merge.

File: extensions/spyre-cli/spyre_cli/core.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        dtype = getattr(torch, tensor_spec.dtype.value)
+        tensor = torch.ones(
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/core.py:50]: Outputs should be torch.empty, not torch.ones — pre-filling with 1.0 lets a launch that never writes the output still print plausible all-ones data, hiding the failure, whereas empty surfaces it as obvious garbage. Related: the output dtype/shape come straight from io.json with no check against the bundle, so a wrong dtype (e.g. the example's float32 output vs a float16 program) silently reinterprets the bytes on readback — worth guarding once we have a signature to validate against.

File: extensions/spyre-cli/spyre_cli/core.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    spyrecode_dir = path / "spyreCodeDir"
+    if not spyrecode_dir.exists():
+        raise FileNotFoundError(
+            f"spyreCodeDir not found at '{spyrecode_dir}'. "
+            "Ensure the model has been compiled before launching."
+        )
+
+    # this is the magic line
+    # should have already compiled at this point
+    jobplan = torch.spyre.prepare_kernel(str(spyrecode_dir))
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/core.py:32]: This prepare+launch sequence (plus the /spyreCodeDir convention) already exists as SpyreSDSCKernelRunner. Should we just delegate to it?

File: extensions/spyre-cli/spyre_cli/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from .core import launch
+
+__all__ = [launch]
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/__init__.py]: I believe items in __all__ should be a string, this is a function

File: extensions/spyre-cli/spyre_cli/iospec/iospec.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+from typing import List
+
+
+class Dtype(str, Enum):
+    FLOAT16 = "float16"
+    FLOAT32 = "float32"
+
+
+class TensorSpec(BaseModel):
+    ndims: int
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/iospec/iospec.py]: do we need ndims? It seems redundant with len(dimensions). It's possible a user may set this to some other value from len(dimensions) as well which could cause issues?

File: extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json:1]: Is this used anywhere?

File: extensions/spyre-cli/spyre_cli/core.py
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/core.py:1]: we should add the license header to all source files

File: extensions/spyre-cli/spyre_cli/core.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    torch.spyre.launch_jobplan(jobplan, tensors)
+
+
+def launch_from_iofile(path=".", iofile="io.json"):
+    path = Path(path)
+    try:
+        spec = parse_json_file(iofile)
+    except Exception as e:
+        print(e)
+        exit(1)
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/core.py]: rather than exit, should we not just raise an exception?

File: extensions/spyre-cli/spyre_cli/core.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        tensor = torch.ones(
+            tensor_spec.dimensions,
+            dtype=dtype,
+            device="spyre",
+        )
+        tensors.append(tensor)
+
+    _launch(path, tensors)
+
+    print(tensors[-1])
```
Comment (JRosenkranz) [inline extensions/spyre-cli/spyre_cli/core.py:74]: nit: this assumes a non-empty output list and that the last tensor is the output

File: extensions/spyre-cli/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,21 @@
+[project]
+name = "spyre-cli"
+version = "0.1.0"
+description = "Add your description here"
```
Comment (JRosenkranz) [inline extensions/spyre-cli/pyproject.toml]: This should be updated

File: extensions/spyre-cli/spyre_cli/core.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    # this is the magic line
+    # should have already compiled at this point
+    jobplan = torch.spyre.prepare_kernel(str(spyrecode_dir))
+    torch.spyre.launch_jobplan(jobplan, tensors)
+
+
+def launch_from_iofile(path=".", iofile="io.json"):
+    path = Path(path)
+    try:
+        spec = parse_json_file(iofile)
```
Comment (fabianlim) [inline extensions/spyre-cli/spyre_cli/core.py:41]: my main concern that who is going to write this iofile, as the compiler does not produce such a file automatically

The actualy use case is probably when the spyurecode was produced, the hope is that an `iofile` is produced along with it, so there is provenance for what tensor sizes it is compiled for.

However reality is different today, as there is no such file. Hence i would imagine that today, users will be forced to create an iofile just to launch, 
- if you cannot remember what sizes it was created before, then you cant launch it anyway. However if you already forgot for what sizes or func that kernel was compiled for, then you cant even create your tensors on device.
- otherwise, you just created the spyrecode, and after creating your tensors on device, you are forced to do one more extra step to create this file.

As such I feel that this file doesnt serve a real purpose today. It woujld however be really useful if such information was recorded at the instance when the spyrecode was created.

File: extensions/spyre-cli/spyre_cli/iospec/iospec.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    dimensions: List[int]  # has to be opened-up for symbolic
+    dtype: Dtype
+
+
+class IOSpec(BaseModel):
```
Comment (fabianlim) [inline extensions/spyre-cli/spyre_cli/iospec/iospec.py:33]: If you are going to create an IOSpec, why dont we create an LoopSpec/OpSpec (that could potentially include nests) that includes
- inputs
- outputs
- function that was compiled, etc..


Actually on retrospec, the Opspec lab seems like a good place to interface with this cli
- if the OpSpec lab allows users to run from OpSpec, and since compilation in torch-spyre follows from OpSpec, then cant we create some metadata in the compilation flow from torch-spyre. cc: @willmj 


https://github.com/torch-spyre/torch-spyre/pull/4060

File: extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json (hunk 1 above)
Comment (ChanderG) [reply extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json:4]: We can move to using symbols in the dimensions list, of the form "x", "y" etc, where the symbol name can also correlate across the inputs, for eg, input 1: ["x", "y"] and input 2: ["y" , "z"] etc.

File: extensions/spyre-cli/spyre_cli/core.py (hunk 10 above)
Comment (ChanderG) [reply extensions/spyre-cli/spyre_cli/core.py:41]: Agreed. We should have the io spec created at compile time - during both the SDSC and the KTIR workflows. That would the be the right thing.

This approach is a stepping stone towards that.

**Comment by mudhakar** [comment]
Hand-written io spec is a problem. 

The io spec should be generated along the primary compile pathway (Inductor, Triton), not authored by the user. 

- io spec is an accompanying JSON to spyre code; io spec + spyre code is standalone launchable
- Schema and data structures for io spec is common across all pathways (inductor, triton)
- Generating io spec is each pathway's responsibility
- io spec should capture external input/output count and order, shapes and dtypes, symbolic HBM addresses, and symbolic (dynamic) shapes with their bounds+granularity.

Cc: @ChanderG , @JRosenkranz , @fabianlim

File: extensions/spyre-cli/pyproject.toml (hunk 9 above)
Comment (dushyantbehl) [reply extensions/spyre-cli/pyproject.toml]: sure @JRosenkranz we will update this

File: extensions/spyre-cli/README.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
+# spyre-cli
+
+## Install
+
+Needs a pod with entire Spyre stack setup.
```
Comment (dushyantbehl) [inline extensions/spyre-cli/README.md]: nit: `Needs a working spyre stack environment`

File: extensions/spyre-cli/spyre_cli/__init__.py (hunk 5 above)
Comment (dushyantbehl) [reply extensions/spyre-cli/spyre_cli/__init__.py]: Agree @JRosenkranz 

nit: `__all__ = ["launch"]`

**Comment by ChanderG** [comment]
@JRosenkranz Have addressed most of your comments. Left it as a series of commits for the moment - so that your previous review comments are not lost. Once approved - I will squash and force push.

There is one thing not done: the last commit tries to use the job_plan struct to validate number of inputs like you had asked, but it looks like that info is not yet populated in the job_plan?

File: extensions/spyre-cli/README.md (hunk 12 above)
Comment (ChanderG) [reply extensions/spyre-cli/README.md]: Done.

File: extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json
Comment (ChanderG) [reply extensions/spyre-cli/spyre_cli/iospec/e_incorrect.json:1]: Meant to be a demonstrative wrong example. Have added a test that uses the file.

File: extensions/spyre-cli/spyre_cli/iospec/iospec.py (hunk 6 above)
Comment (ChanderG) [reply extensions/spyre-cli/spyre_cli/iospec/iospec.py]: Removed it.

File: extensions/spyre-cli/spyre_cli/core.py (hunk 4 above)
Comment (ChanderG) [reply extensions/spyre-cli/spyre_cli/core.py:32]: One advantage keeping it separate - we can use the job plan validation like you mention.

File: extensions/spyre-cli/spyre_cli/core.py (hunk 3 above)
Comment (ChanderG) [reply extensions/spyre-cli/spyre_cli/core.py:50]: Done.

**Review by JRosenkranz — CHANGES_REQUESTED** [review]
Let's split the IOSpec out into a follow-up PR and keep this one to the SDK/launch path. The format is still likely to change — the signature really wants to come from whatever produces the SpyreCode (and be versioned with it), not a hand-authored file — so I'd rather not introduce it as a contract until we actually need it. Happy to merge the launch path now and take up the IOSpec separately once we've settled where it's produced.

Please see https://github.com/torch-spyre/torch-spyre/issues/4210

cc: @fabianlim @mudhakar
